Feature: Add 1.4.0 changelog, update Gnosis RPC endpoint (#549)
safe-global/safe-contracts: CHANGELOG.md

This changelog only contains changes starting from version 1.3.0

# Version 1.4.0

## Compiler settings

> Solidity compiler: [0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6) (more info see issue [#251](https://github.com/safe-global/safe-contracts/issues/251))

Reviewer: ```suggestion
Solidity compiler: [0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6) (for more info see issue [#251](https://github.com/safe-global/safe-contracts/issues/251))
```

-   `CreateCall` at `0x9b35Af71d77eaf8d7e40252370304687390A1A52`
-   `SignMessageLib` at `0x58FCe385Ed16beB4BCE49c8DF34c7d6975807520`

### Storage reader contracts

-   `SimulateTxAccessor` at `0x3d4BA2E0884aa488718476ca2FB8Efc291A46199`

## Changes

### General

#### Drop "Gnosis" from contract names

Removed the "Gnosis" prefix from all contract names.

### Core contract

File: [`contracts/SafeL2.sol`](https://github.com/safe-global/safe-contracts/blob/3c3fc80f7f9aef1d39aaae2b53db5f4490051b0d/contracts/SafeL2.sol)

#### Remove usage of the `GAS` opcode in module execute flows

Issue: [#459](https://github.com/safe-global/safe-contracts/issues/459)

The following rule of usage of the `GAS` opcode in the ERC-4337 standard made it impossible to build a module to support ERC4337:

> -   Must not use GAS opcode (unless followed immediately by one of { CALL, DELEGATECALL, CALLCODE, STATICCALL }.)

We removed the `GAS` opcode usage in module transactions to forward all the available gas instead.

#### Require the `to` address to be a contract in `setupModules`

Issue: [#483](https://github.com/safe-global/safe-contracts/issues/483)

> The `setupModules` method was changed to require the `to` address to be a contract. If the `to` address is not a contract, the transaction will be reverted with a `GS002` error code.

Reviewer: ```suggestion
The `setupModules` method was changed to require the `to` address to be a contract. If the `to` address is not a contract, the transaction will revert with a `GS002` error code.
```

#### Enforce the `dataHash` is equal to `data` in the signature verification process for contract signatures

Issue: [#497](https://github.com/safe-global/safe-contracts/issues/497)

> To prevent unexpected behaviour, the `dataHash` must now equal a hash of the `data` in the signature verification process for contract signatures. Otherwise, the transaction will be reverted with a `GS027` error code.

Reviewer: ```suggestion
To prevent unexpected behaviour, the `dataHash` must now equal a hash of the `data` in the signature verification process for contract signatures. Otherwise, the transaction will revert with a `GS027` error code.
```

#### Check the EIP-165 signature of the Guard before adding

Issue: [#309](https://github.com/safe-global/safe-contracts/issues/309)

When setting a guard, the core contract will check that the target address supports the Guard interface with an EIP-165 check. If it doesn't, the transaction will revert with the `GS300` error code.

#### Index essential parameters when emitting events

Issue: [#541](https://github.com/safe-global/safe-contracts/issues/541)

Index essential parameters in the essential events, such as:

-   Owner additions and removals (Indexed parameter - owner address)
-   Fallback manager changes (Indexed parameter - fallback manager address)
-   Module additions and removals (Indexed parameter - module address)
-   Transaction guard changes (Indexed parameter - guard address)
-   Transaction execution/failure (Indexed parameter - transaction hash)

### Factory

Umbrella issue: [#462](https://github.com/safe-global/safe-contracts/issues/462)

#### Remove the `createProxy` method

This method uses the `CREATE` opcode, which is not counterfactual for a specific deployment. This caused user errors and lost/stuck funds and is now removed.

> #### Add a check that singleton exists for the initializer call

Reviewer: ```suggestion
#### Add a check that Singleton exists for the initializer call
```

> If the initializer data is provided, the factory now checks that the singleton contract exists and the success of the call to avoid a proxy being deployed uninitialized

Reviewer: ```suggestion
If the initializer data is provided, the Factory now checks that the Singleton contract exists and the success of the call to avoid a proxy being deployed uninitialized
```

#### Add `createNetworkSpecificProxy`

> This method will use the chain id in the `CREATE2` salt; therefore, deploying a proxy to the same address on other networks will be impossible.

Reviewer: ```suggestion
This method will use the chain id in the `CREATE2` salt; therefore, deploying a proxy to the same address on other networks is impossible.
```

This method should enable the creation of proxies that should exist only on one network (e.g. specific governance or admin accounts)

#### Remove the `calculateProxyAddress` method

> Methods using the revert approach to return data only work well with some nodes, as they all return messages differently. Hence, we removed it, and the off-chain CREATE2 calculation is still possible.

Reviewer: I _think_ I understood this correctly.

```suggestion
Method uses the revert approach to return data that only works well with some nodes, as they all return messages differently. Hence, we removed it, and the off-chain CREATE2 calculation is still possible.
```
